sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -104,6 +110,17 @@ class QueryExecution(
     ReuseExchange(sparkSession.sessionState.conf),
     ReuseSubquery(sparkSession.sessionState.conf))
 
+  // With adaptive execution, whole stage codegen will be done inside `QueryStageExecutor`.
+  protected def adaptivePreparations: Seq[Rule[SparkPlan]] = Seq(
```

Comment:
More of a high level note, we really should consider replacing this with a RuleExecutor.